lib/grammers-client/examples/downloader.rs

```diff
@@ -0,0 +1,195 @@
+
+async fn async_main() -> Result<()> {
+    SimpleLogger::new()
+        .with_level(log::LevelFilter::Info)
+        .init()
+        .unwrap();
+
+    let api_id = env!("TG_ID").parse().expect("TG_ID invalid");
+    let api_hash = env!("TG_HASH").to_string();
+    let chat_name = env::args().skip(1).next().expect("chat name missing");
+
+    println!("Connecting to Telegram...");
+    let mut client = Client::connect(Config {
+        session: Session::load_file_or_create(SESSION_FILE)?,
+        api_id,
+        api_hash: api_hash.clone(),
+        params: Default::default(),
+    })
+    .await?;
+    println!("Connected!");
+
+    // If we can't save the session, sign out once we're done.
+    let mut sign_out = false;
+
+    if !client.is_authorized().await? {
+        println!("Signing in...");
+        let phone = prompt("Enter your phone number (international format): ")?;
+        let token = client.request_login_code(&phone, api_id, &api_hash).await?;
+        let code = prompt("Enter the code you received: ")?;
+        let signed_in = client.sign_in(&token, &code).await;
+        match signed_in {
+            Err(SignInError::PasswordRequired(password_token)) => {
+                // Note: this `prompt` method will echo the password in the console.
+                //       Real code might want to use a better way to handle this.
+                let hint = password_token.hint().unwrap();
+                let prompt_message = format!("Enter the password (hint {}): ", &hint);
+                let password = prompt(prompt_message.as_str())?;
+
+                client
+                    .check_password(password_token, password.trim())
+                    .await?;
+            }
+            Ok(_) => (),
+            Err(e) => panic!("{}", e),
+        };
+        println!("Signed in!");
+        match client.session().save_to_file(SESSION_FILE) {
+            Ok(_) => {}
+            Err(e) => {
+                println!(
+                    "NOTE: failed to save the session, will sign out when done: {}",
+                    e
+                );
+                sign_out = true;
+            }
+        }
+    }
+
+    // Obtain a `ClientHandle` to perform remote calls while `Client` drives the connection.
+    //
+    // This handle can be `clone()`'d around and freely moved into other tasks, so you can invoke
+    // methods concurrently if you need to. While you do this, the single owned `client` is the
+    // one that communicates with the network.
+    //
+    // The design's annoying to use for trivial sequential tasks, but is otherwise scalable.
+    let mut client_handle = client.clone();
+    let network_handle = task::spawn(async move { client.run_until_disconnected().await });
+
+    let chat = client_handle.resolve_username(chat_name.as_str()).await?;
+
+    if chat.is_none() {
+        panic!("Chat {} could not be found", chat_name);
+    }
+
+    let mut messages = client_handle.iter_messages(&chat.unwrap());
```

Comment:
`is_none` followed by `unwrap` is not very idiomatic. We could change it with `.expect("Chat could not be found")` or maybe a `match`.